_drafts/diving-into-jtag-part6.md

```diff
@@ -0,0 +1,373 @@
+   read `IDCODE` and with some luck you will be able to determine not only the JTAG
+   pins, but also the type of microcontroller. If the `TDO` pin reads only ones or
+   zeros, it means that you have incorrectly assigned the pins in step 1 and you need
+   to reassign the pins and repeat the whole procedure again.
+
+However, this method has a number of disadvantages:
+
+- This method does not find the `TDI` pin.
+- If the JTAG circuit contains more than one TAP, it may take more than 32
+  iterations to find `IDCODE`, which means that if after e.g. 64 iterations
+  you do not get a `TDO` pattern similar to `IDCODE` on the `TDO` pin, it
+  does not mean that you have selected the wrong pins, it is possible that
+  the JTAG circuit contains many TAPs.
+
+### BYPASS Scan
+
+This method uses the `BYPASS` register to scan the JTAG circuit. The `BYPASS`
+register allows you to skip a logic element in the circuit and directly
+transfer data from input to output, minimizing latency.
+
+1. The first step is similar to the first step in the previous method.
+2. Send the sequence `0b01100` on the selected `TMS` pin to move thus to the
+   `Shift-IR` state.
+3. Now it is necessary to send the `BYPASS` command, but since the size of the
+   `IR` register is not known, but it is necessary to send a large number of
+   units (as many as 1024, for example).
+4. Send the sequence `0b11000` to the selected `TMS` pin, thus moving to the
+   `Shift-DR` state.
+5. Using the `TDI` output - send some bit pattern that can be easily recognized.
+6. Read the bits from the `TDO` pin and if a bit pattern is detected, it means
+   that you have selected the pins correctly.
+
+This algorithm is slower than the previous one, but it can detect all JTAG pins.
+
+>If you have a lot of candidate output, it is recommended to first reduce the number
+of candidates using `IDCODE Scan` and then refine `BYPASS Scan`.
+
+Of course every time to do all these steps manually is quite boring and tedious so
+smart people have long ago automated this process. Quite popular tools are:
+[JTAGenum](https://github.com/cyphunk/JTAGenum/tree/master) and
+[JTAGULATOR](http://www.grandideastudio.com/jtagulator/).
+
+>If the highest level of RDP 2 protection has been applied to the microcontroller,
+JTAGulator and similar tools will not help.
+
+Examples of such attacks are described in more detail in the following sources:
+
+- [JTAG Scanning](https://github.com/cyphunk/JTAGenum/wiki/Embedded-Analysis#jtag-scanning)
+- [Hardware Debugging for Reverse Engineers Part 2: JTAG, SSDs and Firmware Extraction](https://wrongbaud.github.io/posts/jtag-hdd/)
+- [JTAGulator: Introduction and Demonstration](https://youtu.be/uVIsbXzQOIU)
+- [03 - How To Find The JTAG Interface - Hardware Hacking Tutorial](https://youtu.be/_FSM_10JXsM)
+
+## Attack on JTAG
+
+The next thing to determine is the number of devices in the JTAG chain, because even
```

Comment:
Similar phrasing suggestion here:
```suggestion
The next thing an attacker will determine is the number of devices in the JTAG chain, because even
```

Reply:
Oops 2 :)